lib/aws/ddb/internal.ts

```diff
@@ -26,10 +26,22 @@ async function main (method: "transactGetItems", params: DynamoDB.TransactGetIte
 async function main (method: "transactWriteItems", params: DynamoDB.TransactWriteItemsInput): Promise<DynamoDB.TransactWriteItemsOutput>;
 
 async function main (method: string, params: any): Promise<any> {
-	log({"level": "debug", "category": `aws:dynamodb:${method}:request`, "message": JSON.stringify(params, null, 4), "payload": {"request": params}});
-	const result = await ddb()[method](params).promise();
-	log({"level": "debug", "category": `aws:dynamodb:${method}:response`, "message": typeof result === "undefined" ? "undefined" : JSON.stringify(result, null, 4), "payload": {"response": result}});
-	return result;
+	const func = ddb()[method](params);
+    func.on("extractError", ({ error, httpResponse }) => {
+        try {
+          if (error) {
+            const { CancellationReasons } = JSON.parse(httpResponse.body.toString());
+            error.cancellationReasons = CancellationReasons;
+          }
+        } catch (err) {
+            throw err;
+        }
+    });
```

Comment:
First of all **LOVE** how this is contained in only one place in the code. That is really awesome. When I first thought about this issue, I didn't think to put it in here. Really clever, and **really love it**. So great job!

However, I can't decide if I'm a fan of having this get called on every request made to AWS. It seems like it adds a lot of overhead and could impose performance risks.

That being said, I'm not sure of a better solution. Ideally we would implement this functionality as a plugin. And the plugin would handle it, so that it isn't being run on every request. But plugin support isn't yet available in Dynamoose. So not really sure how to handle this.

Do you have any thoughts?

Reply:
@KennethWKZ Got it. Interesting that aws-sdk v3 fixed this. Makes me question if we should even merge this in. aws-sdk v3 is already implemented in Dynamoose v3 alpha.

I'm honestly unsure if v2 is going to have any more releases. It might, but just unsure at this point. Really wanting to shift focus to v3.

I'm going to keep this PR open for now, and if we do end up making another v2 release, we can merge it in. Otherwise, I'd suggest using your own fork in the meantime.

Reply:
ok thanks! @fishcharlie